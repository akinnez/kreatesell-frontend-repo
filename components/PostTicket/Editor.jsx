import React from "react";
import dynamic from "next/dynamic";
const ReactQuill = dynamic(() => import("react-quill"), { ssr: false });
import EditorToolbar, { modules, formats } from "./EditorToolbar";
import "react-quill/dist/quill.snow.css";
import "./editor.module.scss";

export const Editor = () => {
  const [state, setState] = React.useState({ value: null });
  const handleChange = (value) => {
    setState({ value });
  };
  return (
    <div className="text-editor">
      <EditorToolbar />
      <ReactQuill
        theme="snow"
        value={state.value}
        onChange={handleChange}
        placeholder={"Write a reply message here..."}
        modules={modules}
        formats={formats}
      />
    </div>
  );
};

export default Editor;
