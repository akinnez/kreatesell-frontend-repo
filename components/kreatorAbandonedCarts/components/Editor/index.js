import { useState, useMemo, useRef } from "react";
import dynamic from "next/dynamic";
import { Typography } from "antd";
import Toolbar from "../Toolbar";
import PersonificationModal from "../PersonificationModal";
import "react-quill/dist/quill.snow.css";
import styles from "./index.module.scss";

const { Link: AntLink } = Typography;

const ReactQuill = dynamic(
  async function () {
    const { default: RQ } = await import("react-quill");
    return function MyRQ({ forwardedRef, ...props }) {
      return <RQ ref={forwardedRef} {...props} />;
    };
  },
  { ssr: false }
);

const Editor = ({ formik }) => {
  const [personificationModal, setPersonificationModal] = useState(false);
  const editorElement = useRef(null);
  const editorContainer = useRef(null);

  const modules = useMemo(() => {
    return {
      toolbar: {
        container: "#toolbar",
        handlers: {
          personification: () => setPersonificationModal(true),
        },
      },
    };
  }, []);

  const handleLabelEnter = () => {
    if (!formik.touched.email_content && !formik.errors.email_content) {
      editorContainer.current.classList.add(styles["label--hover"]);
    }
  };

  const handleLabelLeave = () => {
    editorContainer.current.classList.remove(styles["label--hover"]);
  };

  const handleLabelClick = () => {
    editorElement.current.focus();
    editorContainer.current.classList.remove(styles["label--hover"]);
  };

  const handleEditorChange = value => {
    const result = value === "<p><br></p>" ? "" : value;
    formik.setFieldValue("email_content", result);
  };

  const handleEditorBlur = () => {
    if (!formik.values.email_content) {
      formik.setFieldError("email_content", "Enter email content");
    }

    formik.setFieldTouched("email_content");
  };

  const handlePersonification = personifications => {
    const element = editorElement.current;
    const start = element.getEditorSelection()?.index ?? 0;
    const length = element.getEditorSelection()?.length ?? 0;

    if (length > 0) {
      element.getEditor().deleteText(start, length);
    }

    element.getEditor().insertText(start, " ");
    element.getEditor().insertText(start + 1, personifications, {
      bold: true,
      italic: true,
      underline: true,
    });

    // const end = element.editor.getLength() - 1;
    // element.getEditor().insertText(end, " ", {
    //   bold: false,
    //   italic: false,
    //   underline: false,
    // });
  };

  return (
    <>
      <div className={styles.editor}>
        <div>
          <label
            onMouseEnter={handleLabelEnter}
            onMouseLeave={handleLabelLeave}
            onClick={handleLabelClick}
          >
            Email Content
          </label>
          <AntLink>Add Media</AntLink>
        </div>
        <div
          ref={editorContainer}
          className={
            formik.touched.email_content && formik.errors.email_content
              ? `${styles.editor__container} ${styles["editor__container--error"]}`
              : `${styles.editor__container}`
          }
        >
          <Toolbar />
          <ReactQuill
            forwardedRef={editorElement}
            modules={modules}
            placeholder="Email Content"
            onChange={handleEditorChange}
            onBlur={handleEditorBlur}
            value={formik.values.email_content}
          />
        </div>
        <div
          className={
            formik.touched.email_content && formik.errors.email_content
              ? `${styles["editor__error-msg"]} ${styles["editor__error-msg--active"]}`
              : `${styles["editor__error-msg"]}`
          }
        >
          {formik.touched.email_content && formik.errors.email_content}
        </div>
      </div>
      {personificationModal && (
        <PersonificationModal
          personificationModal={personificationModal}
          setPersonificationModal={setPersonificationModal}
          handlePersonification={handlePersonification}
        />
      )}
    </>
  );
};

export default Editor;
