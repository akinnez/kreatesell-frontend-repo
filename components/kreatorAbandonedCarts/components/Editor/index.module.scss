@use "public/css/colors";
@use "public/css/mixins";
@use "public/css/screens";

.editor {
  margin-top: 35px;
  margin-bottom: 35px;

  &:hover {
    .editor__container {
      border-color: #40a9ff;
    }

    .editor__container--error {
      border-color: #ff4d4f;

      &:focus-within {
        border-color: #ff7875;
      }
    }
  }

  & > div:first-child {
    @include mixins.flex(row, space-between, center);
    margin-bottom: 10px;

    label {
      font-weight: 500;
      font-size: 16px;
      color: colors.$black-100;
    }

    a {
      font-weight: 500;
      font-size: 16px;
      color: colors.$base-blue;
    }
  }
}

.editor__container {
  border-radius: 8px;
  border: 1px solid #d9d9d9;
  transition: all 0.3s;

  &:focus-within {
    border-color: #40a9ff;
    box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
  }

  > div[id="toolbar"] {
    outline: none;
    border: none;
    border-radius: 8px 8px 0 0;
    box-shadow: 0 1px 10px rgba(0, 0, 0, 0.05);
    @include mixins.flex(row, flex-start, center);
    flex-wrap: wrap;
    gap: 5px 10px;
  }

  div[class^="ql-container"] {
    border-radius: 0 0 8px 8px;
    border: none;
    outline: none;

    > [class^="ql-editor"] {
      min-height: 355px;
      padding: 6.5px 11px;
      outline: none;
      font-size: 16px;

      &::before {
        font-style: normal;
        color: #d9d9d9;
        left: 11px;
      }
    }
  }
}

.editor__container--error {
  border: 1px solid #ff4d4f;

  &:focus-within {
    border-color: #ff7875;
    box-shadow: 0 0 0 2px rgba(255, 77, 79, 0.2);
  }
}

.label--hover {
  border-color: #40a9ff;
}

.editor__error-msg {
  color: #ff4d4f;
  height: 0;
  margin-bottom: 25px;
  transition: height 0.3s linear;
}

.editor__error-msg--active {
  height: 22px;
}
