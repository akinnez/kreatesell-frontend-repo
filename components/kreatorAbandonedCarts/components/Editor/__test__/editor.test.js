import Editor from '..';
import renderer from 'react-test-renderer';

const mockProp = {
	formik: {
		values: {
			email_content: '',
		},
		setFieldTouched: jest.fn(),
		setFieldValue: jest.fn(),
	},
};

it('renders a snapshot of <Editor />', () => {
	const editor = renderer.create(<Editor {...mockProp} />).toJSON();
	expect(editor).toMatchSnapshot();
});
