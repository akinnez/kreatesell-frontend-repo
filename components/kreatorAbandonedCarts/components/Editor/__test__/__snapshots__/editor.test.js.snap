// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot of <Editor /> 1`] = `
<div
  className="editorWrapper"
>
  <div>
    <label
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Email Content
    </label>
    <a
      className="ant-typography"
      onClick={null}
      style={
        Object {
          "WebkitLineClamp": undefined,
        }
      }
    >
      Add Media
    </a>
  </div>
  <div
    className="editorContainer"
  >
    <div
      id="toolbar"
    >
      <select
        className="ql-header"
        defaultValue=""
        onChange={[Function]}
      >
        <option
          value="1"
        />
        <option
          value="2"
        />
        <option
          value="3"
        />
        <option
          value=""
        />
      </select>
      <button
        className="ql-bold"
      />
      <button
        className="ql-italic"
      />
      <button
        className="ql-underline"
      />
      <button
        className="ql-strike"
      />
      <select
        className="ql-align"
        defaultValue=""
        onChange={[Function]}
      >
        <option
          value=""
        />
        <option
          value="center"
        />
        <option
          value="right"
        />
      </select>
      <button
        className="ql-list"
        value="bullet"
      />
      <button
        className="ql-list"
        value="ordered"
      />
      <select
        className="ql-color"
        onChange={[Function]}
      />
      <button
        className="ql-link"
      />
      <button
        className="ql-image"
      />
      <button
        className="ql-video"
      />
    </div>
  </div>
</div>
`;
